crates/prql-parser/src/error.rs

```diff
@@ -0,0 +1,196 @@
+    source_id: u16,
+) -> Error {
+    // TODO: is there a neater way of taking a span? We want to take it based on
+    // the chars, not the bytes, so can't just index into the str.
+    let found = source
+        .chars()
+        .skip(e.span().start)
+        .take(e.span().end() - e.span().start)
+        .collect();
+    let span = Span {
+        start: e.span().start,
+        end: e.span().end,
+        source_id,
+    };
+
+    Error {
+        span,
+        kind: ErrorKind::Lexer(LexerError(found)),
+    }
+}
+
+pub(crate) fn convert_parser_error(e: PError) -> Error {
+    let mut span = e.span();
+
+    if e.found().is_none() {
+        // found end of file
+        // fix for span outside of source
+        if span.start > 0 && span.end > 0 {
+            span.start -= 1;
+            span.end -= 1;
+        }
+    }
+
+    if let SimpleReason::Custom(message) = e.reason() {
+        return Error {
+            span: *span,
+            kind: ErrorKind::Parser(ParserError(message.clone())),
+        };
+    }
+
+    fn token_to_string(t: Option<Token>) -> String {
+        t.map(|t| DisplayToken(&t).to_string())
+            .unwrap_or_else(|| "end of input".to_string())
+    }
+
+    let is_all_whitespace = e
+        .expected()
+        .all(|t| matches!(t, None | Some(Token::NewLine)));
+    let expected: Vec<String> = e
+        .expected()
+        // TODO: could we collapse this into a `filter_map`? (though semantically
+        // identical)
+        //
+        // Only include whitespace if we're _only_ expecting whitespace
+        .filter(|t| is_all_whitespace || !matches!(t, None | Some(Token::NewLine)))
+        .cloned()
+        .map(token_to_string)
+        .collect();
+
+    let while_parsing = e
+        .label()
+        .map(|l| format!(" while parsing {l}"))
+        .unwrap_or_default();
+
+    if expected.is_empty() || expected.len() > 10 {
+        let label = token_to_string(e.found().cloned());
+
+        return Error {
+            span: *span,
+            kind: ErrorKind::Parser(ParserError(format!("unexpected {label}{while_parsing}"))),
+        };
+    }
+
+    let mut expected = expected;
+    expected.sort();
+
+    let expected = match expected.len() {
+        1 => expected.remove(0),
+        2 => expected.join(" or "),
+        _ => {
+            let last = expected.pop().unwrap();
+            format!("one of {} or {last}", expected.join(", "))
+        }
+    };
+
+    Error {
+        span: *span,
+        kind: ErrorKind::Parser(ParserError(match e.found() {
+            Some(found) => format!(
+                "{who}expected {expected}, but found {found}",
+                who = e.label().map(|l| format!("{l} ")).unwrap_or_default(),
+                found = DisplayToken(found)
+            ),
+
+            // We want a friendlier message than "found end of input"...
+            None => format!("Expected {expected}, but didn't find anything before the end."),
```

Comment:
Are we losing anything with this stringification? Or once we leave `prql-parser`, we're can't do anything with this structured info anyway?

Reply:
I don't think we're loosing anything. If we at some point in the future want to provide more structured parser errors we can just change the `ParserError` struct to be an enum instead.

Reply:
Right now, we are not. All errors eventually get converted into `ErrorMessage`, which contains just `message: String, code: String` (and a few others).